File: 백준/Gold/20168. 골목 대장 호석 － 기능성/골목 대장 호석 － 기능성.py
import sys
import heapq

input = sys.stdin.readline

n, m, start, last, value = map(int, input().split())
graph = [[] for _ in range(n + 1)]
before = [[] for _ in range(n + 1)]

for _ in range(m):
    a1, a2, a3 = map(int, input().split())
    graph[a1].append((a2, a3))
    graph[a2].append((a1, a3))

distance = [sys.maxsize] * (n + 1)


def dijk(start, last):
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        val, ver = heapq.heappop(q)
        if distance[ver] < val:
            continue
        for go, va in graph[ver]:
            if va + val <= value:
                if distance[go] > va + val:
                    distance[go] = va + val
                    before[go].append(ver)
                    if go != last:
                        heapq.heappush(q, (va + val, go))


dijk(start, last)
ans = sys.maxsize
kk = sys.maxsize


def dfs(be, next):
    global kk
    if be == start:
        return
    for a1, a2 in graph[be]:
        if a1 == next:
            kk = max(kk, a2)
            for j in before[a1]:
                dfs(a1, j)
            break

for i in before[last]:
    kk = -sys.maxsize
    dfs(last, i)
    ans = min(ans, kk)
if ans == sys.maxsize:
    print(-1)
else:
    print(ans)

File: 백준/Gold/20168. 골목 대장 호석 － 기능성/골목 대장 호석 － 기능성.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {


    public static class go {
        int ver, weight;

        go(int ver, int weight) {
            this.ver = ver;
            this.weight = weight;
        }
    }

    static List<List<go>> graph;
    static int[] distance;
    static int n;

    public static int dijk(int start, int gone, int c, long check) {

        PriorityQueue<go> q = new PriorityQueue<>((x, y) -> {
            return x.weight - y.weight;
        });
        for (int i = 1; i <= n; i++) {
            distance[i] = Integer.MAX_VALUE;
        }
        distance[start] = 0;
        q.add(new go(start, 0));
        int temp = Integer.MAX_VALUE;

        while (!q.isEmpty()) {
            go a = q.poll();
            if (a.ver == gone) {
                temp = Math.min(temp, distance[gone]);
                continue;
            }
            if (a.weight > distance[a.ver]) continue;

            for (go g : graph.get(a.ver)) {
                int now = a.weight + g.weight;
                if (distance[g.ver] < now || g.weight > check)
                    continue;
                if (now > c)
                    continue;
                distance[g.ver] = now;
                q.add(new go(g.ver, now));
            }
        }
        if (temp == Integer.MAX_VALUE)
            return -1;
        else
            return 1;

    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int m, a, b, c;
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        a = Integer.parseInt(st.nextToken());
        b = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        graph = new ArrayList<>();
        distance = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        long left = 0;
        long right = c + 1;
        long answer = Integer.MAX_VALUE;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            graph.get(from).add(new go(to, weight));
            graph.get(to).add(new go(from, weight));

        }

        while (left + 1 < right) {
            long mid = (left + right) / 2;
            int result = dijk(a, b, c, mid);
            if (result == 1) {
                right = mid;
                answer = mid;
            } else {
                left = mid;
            }
        }
        if (answer == Integer.MAX_VALUE)
            System.out.println(-1);
        else
            System.out.println(answer);


    }

}
